src/node_dependencies/traverse.js

```diff
@@ -0,0 +1,142 @@
+const { dirname } = require('path')
+
+const { getModuleName } = require('./module')
+const { getNestedDependencies, handleModuleNotFound } = require('./nested')
+const { getPublishedFiles } = require('./published')
+const { resolvePackage } = require('./resolve')
+const { getSideFiles } = require('./side_files')
+
+const EXCLUDED_MODULES = new Set(['aws-sdk'])
+
+// Local cache used for optimizing the traversal of module dependencies.
+const getNewCache = () => ({ localFiles: new Set(), moduleNames: new Set(), modulePaths: new Set() })
+
+// When a file requires a module, we find its path inside `node_modules` and
+// use all its published files. We also recurse on the module's dependencies.
+const getDependencyPathsForDependency = async function ({
+  dependency,
+  basedir,
+  state,
+  packageJson,
+  pluginsModulesPath,
+}) {
+  const moduleName = getModuleName(dependency)
+
+  // Happens when doing require("@scope") (not "@scope/name") or other oddities
+  // Ignore those.
+  if (moduleName === null) {
+    return []
+  }
+
+  try {
+    return await getDependenciesForModuleName({ moduleName, basedir, state, pluginsModulesPath })
+  } catch (error) {
+    return handleModuleNotFound({ error, moduleName, packageJson })
+  }
+}
+
+const getDependencyNamesAndPathsForDependencies = async function ({
+  dependencies: dependencyNames,
+  basedir,
+  state = getNewCache(),
+  packageJson,
+  pluginsModulesPath,
+}) {
+  const dependencies = await Promise.all(
+    dependencyNames.map((dependencyName) =>
+      getDependencyNamesAndPathsForDependency({
+        dependency: dependencyName,
+        basedir,
+        state,
+        packageJson,
+        pluginsModulesPath,
+      }),
+    ),
+  )
+  const { moduleNames, paths } = dependencies.reduce(
+    (aggregate, dependency) => {
+      dependency.moduleNames.forEach(aggregate.moduleNames.add, aggregate.moduleNames)
+      dependency.paths.forEach(aggregate.paths.add, aggregate.paths)
+
+      return aggregate
+    },
+    { moduleNames: new Set(), paths: new Set() },
+  )
```

Comment:
[sand] Would the following lines be simpler? (feel free to discard this comment)

```js
const moduleNames = new Set(dependencies.map(dependency => dependency.moduleNames))
const paths = new Set(dependencies.map(dependency => dependency.paths))
```

Reply:
Absolutely. Changed in https://github.com/netlify/zip-it-and-ship-it/pull/325/commits/3f0459cea8a2f2bff47faf090aaa638374c6a019.